packages/i18n/I18nDecorator/I18nDecorator.js

```diff
@@ -66,49 +91,97 @@ const I18nDecorator = hoc((config, Wrapped) => {	// eslint-disable-line no-unuse
 		constructor (props) {
 			super(props);
 			const ilibLocale = ilib.getLocale();
-			const locale = props.locale && props.locale !== ilibLocale ? updateLocale(props.locale) : ilibLocale;
+			const locale = props.locale && props.locale !== ilibLocale ? props.locale : ilibLocale;
 
-			this.state = {
-				locale: locale,
-				rtl: isRtlLocale(),
-				updateLocale: this.updateLocale
-			};
+			this.loadResourceJob = new Job(this.setState.bind(this));
+			this.state = this.getDerivedStateForLocale(locale);
 		}
 
 		getChildContext () {
 			return {
-				Subscriber: this.publisher.getSubscriber(),
-				rtl: isRtlLocale(),
+				rtl: this.state.rtl,
 				updateLocale: this.updateLocale
 			};
 		}
 
-		componentWillMount () {
+		componentDidMount () {
 			this.publisher = Publisher.create('i18n', this.context.Subscriber);
 			this.publisher.publish({
 				locale: this.state.locale,
-				rtl: isRtlLocale()
+				rtl: this.state.rtl
 			});
-		}
 
-		componentDidMount () {
 			if (typeof window === 'object') {
 				on('languagechange', this.handleLocaleChange, window);
 			}
+
+			this.loadResources(this.state.locale);
 		}
 
-		componentWillReceiveProps (newProps) {
-			if (newProps.locale) {
-				this.updateLocale(newProps.locale);
+		componentWillReceiveProps (nextProps) {
+			if (this.props.locale !== nextProps.locale) {
+				const state = this.getDerivedStateForLocale(nextProps.locale);
+				if (sync) {
+					this.setState(state);
+				} else {
+					this.loadResources(nextProps.locale);
+				}
 			}
 		}
 
 		componentWillUnmount () {
+			this.loadResourceJob.stop();
 			if (typeof window === 'object') {
 				off('languagechange', this.handleLocaleChange, window);
 			}
 		}
 
+		getDerivedStateForLocale (spec) {
+			const locale = updateLocale(spec);
+
+			let rtl = false;
```

Comment:
Is there any reason to default to `false` while waiting for a resource load?  Could this cause thrashing when switching from one rtl to another rtl locale?

Reply:
Good call. I'll rework this to only update that state member when in sync.